File: Example/ExampleObjC/ExampleCropViewController.m
//
//  ExampleCropViewController.m
//  Example
//
//  Created by Vitalii Parovishnyk on 2/11/17.
//  Copyright © 2017 IGR Software. All rights reserved.
//

@import UIKit;

#import "ExampleCropViewController.h"

@interface ExampleCropViewController ()

@property (nonatomic, weak) IBOutlet UISlider *angleSlider;
@property (nonatomic, weak) IBOutlet UILabel *angleLabel;

@end

@implementation ExampleCropViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    //FIXME: Themes Preview
    //IGRCropLine.appearance().backgroundColor = UIColor.green
    //IGRCropGridLine.appearance().backgroundColor = UIColor.yellow
    //IGRCropCornerView.appearance().backgroundColor = UIColor.purple
    //IGRCropCornerLine.appearance().backgroundColor = UIColor.orange
    //IGRCropMaskView.appearance().backgroundColor = UIColor.blue
    //IGRPhotoContentView.appearance().backgroundColor = UIColor.gray
    //IGRPhotoTweakView.appearance().backgroundColor = UIColor.brown
    
    //FIXME: Customization
    //self.photoView.isHighlightMask = true
    //self.photoView.highlightMaskAlphaValue = 0.2
    //self.maxRotationAngle = CGFloat(M_PI)
    //self.isAutoSaveToLibray = true
    
    [self setupSlider];
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

- (void)setupSlider {
    self.angleSlider.minimumValue = -[IGRRadianAngle toRadians:45.0];
    self.angleSlider.maximumValue = [IGRRadianAngle toRadians:45.0];
    self.angleSlider.value = CGFloat.zero;
    
    [self setupAngleLabelValue:self.angleSlider.value];
}

- (void)setupAngleLabelValue:(CGFloat)radians {
    NSUInteger intDegrees = [IGRRadianAngle toDegrees:radians];
    self.angleLabel.text = [NSString stringWithFormat:@"%@°", @(intDegrees)];
}

// MARK: - Actions

- (IBAction)onChandeAngleSliderValue:(UISlider *)sender {
    CGFloat radians = sender.value;
    [self setupAngleLabelValue:radians];
    [self changeAngleWithValue:radians];
}

- (IBAction)onEndTouchAngleControl:(UIControl *)sender {
    [self stopChangeAngle];
}

- (IBAction)onChandeAnglePickerViewValue:(UIPickerView *)sender {
    CGFloat radians = CGFloat.zero;
    [self setupAngleLabelValue:radians];
    [self changeAngleWithValue:radians];
}

- (IBAction)onTouchResetButton:(UIButton *)sender {
    self.angleSlider.value = CGFloat.zero;
    [self setupAngleLabelValue:self.angleSlider.value];
    
    [self resetView];
}

- (IBAction)onTouchCancelButton:(UIButton *)sender {
    [self dismissAction];
}

- (IBAction)onTouchCropButton:(UIButton *)sender {
    [self cropAction];
}


- (IBAction)onTouchAspectButton:(UIButton *)sender {
    UIAlertController *actionSheet = [UIAlertController alertControllerWithTitle:nil message:nil preferredStyle:UIAlertControllerStyleActionSheet];
    
    [actionSheet addAction:[UIAlertAction actionWithTitle:@"Original"
                                                    style:UIAlertActionStyleDefault
                                                  handler:^(UIAlertAction * _Nonnull action) {
                                                      [self resetAspectRect];
                                                  }]];
    
    [actionSheet addAction:[UIAlertAction actionWithTitle:@"Squere"
                                                    style:UIAlertActionStyleDefault
                                                  handler:^(UIAlertAction * _Nonnull action) {
                                                      [self setCropAspectRectWithAspect:@"1:1"];
                                                  }]];
    
    [actionSheet addAction:[UIAlertAction actionWithTitle:@"2:3"
                                                    style:UIAlertActionStyleDefault
                                                  handler:^(UIAlertAction * _Nonnull action) {
                                                      [self setCropAspectRectWithAspect:@"2:3"];
                                                  }]];
    
    [actionSheet addAction:[UIAlertAction actionWithTitle:@"3:5"
                                                    style:UIAlertActionStyleDefault
                                                  handler:^(UIAlertAction * _Nonnull action) {
                                                      [self setCropAspectRectWithAspect:@"3:5"];
                                                  }]];
    
    [actionSheet addAction:[UIAlertAction actionWithTitle:@"3:4"
                                                    style:UIAlertActionStyleDefault
                                                  handler:^(UIAlertAction * _Nonnull action) {
                                                      [self setCropAspectRectWithAspect:@"3:4"];
                                                  }]];
    
    [actionSheet addAction:[UIAlertAction actionWithTitle:@"5:7"
                                                    style:UIAlertActionStyleDefault
                                                  handler:^(UIAlertAction * _Nonnull action) {
                                                      [self setCropAspectRectWithAspect:@"5:7"];
                                                  }]];
    
    [actionSheet addAction:[UIAlertAction actionWithTitle:@"9:16"
                                                    style:UIAlertActionStyleDefault
                                                  handler:^(UIAlertAction * _Nonnull action) {
                                                      [self setCropAspectRectWithAspect:@"9:16"];
                                                  }]];
    
    [self presentViewController:actionSheet animated:YES completion:nil];
}

@end

File: Example/ExampleCropViewController.swift
//
//  ExampleCropViewController.swift
//  IGRPhotoTweaks
//
//  Created by Vitalii Parovishnyk on 2/7/17.
//  Copyright © 2017 IGR Software. All rights reserved.
//

import IGRPhotoTweaks

import UIKit
import HorizontalDial

class ExampleCropViewController: IGRPhotoTweakViewController {
    
    /**
     Slider to change angle.
     */
    @IBOutlet weak fileprivate var angleSlider: UISlider?
    @IBOutlet weak fileprivate var angleLabel: UILabel?
    @IBOutlet weak fileprivate var horizontalDial: HorizontalDial? {
        didSet {
            self.horizontalDial?.migneticOption = .none
        }
    }
    
    // MARK: - Life Cicle
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        self.setupSlider()
    }
    
    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }
    
    //FIXME: Themes Preview
//    override open func setupThemes() {
//
//        IGRCropLine.appearance().backgroundColor = UIColor.green
//        IGRCropGridLine.appearance().backgroundColor = UIColor.yellow
//        IGRCropCornerView.appearance().backgroundColor = UIColor.purple
//        IGRCropCornerLine.appearance().backgroundColor = UIColor.orange
//        IGRCropMaskView.appearance().backgroundColor = UIColor.blue
//        IGRPhotoContentView.appearance().backgroundColor = UIColor.gray
//        IGRPhotoTweakView.appearance().backgroundColor = UIColor.brown
//    }
    
    fileprivate func setupSlider() {
        self.angleSlider?.minimumValue = -Float(IGRRadianAngle.toRadians(45))
        self.angleSlider?.maximumValue = Float(IGRRadianAngle.toRadians(45))
        self.angleSlider?.value = 0.0
        
        setupAngleLabelValue(radians: CGFloat((self.angleSlider?.value)!))
    }
    
    fileprivate func setupAngleLabelValue(radians: CGFloat) {
        let intDegrees: Int = Int(IGRRadianAngle.toDegrees(radians))
        self.angleLabel?.text = "\(intDegrees)°"
    }
    
    // MARK: - Rotation
    
    override func willTransition(to newCollection: UITraitCollection, with coordinator: UIViewControllerTransitionCoordinator) {
        super.willTransition(to: newCollection, with: coordinator)
        
        coordinator.animate(alongsideTransition: { (context) in
            self.view.layoutIfNeeded()
        }) { (context) in
            //
        }
    }
    
    // MARK: - Actions
    
    @IBAction func onChandeAngleSliderValue(_ sender: UISlider) {
        let radians: CGFloat = CGFloat(sender.value)
        setupAngleLabelValue(radians: radians)
        self.changeAngle(radians: radians)
    }
    
    @IBAction func onEndTouchAngleControl(_ sender: UIControl) {
        self.stopChangeAngle()
    }
    
    @IBAction func onTouchResetButton(_ sender: UIButton) {
        self.angleSlider?.value = 0.0
        self.horizontalDial?.value = 0.0
        setupAngleLabelValue(radians: 0.0)
        
        self.resetView()
    }
    
    @IBAction func onTouchCancelButton(_ sender: UIButton) {
        self.dismissAction()
    }
    
    @IBAction func onTouchCropButton(_ sender: UIButton) {
        cropAction()
    }
    
    @IBAction func onTouchAspectButton(_ sender: UIButton) {
        let actionSheet = UIAlertController(title: nil,
                                            message: nil,
                                            preferredStyle: .actionSheet)
        
        
        actionSheet.addAction(UIAlertAction(title: "Original", style: .default) { (action) in
            self.resetAspectRect()
        })
        
        actionSheet.addAction(UIAlertAction(title: "Squere", style: .default) { (action) in
            self.setCropAspectRect(aspect: "1:1")
        })
        
        actionSheet.addAction(UIAlertAction(title: "2:3", style: .default) { (action) in
            self.setCropAspectRect(aspect: "2:3")
        })
        
        actionSheet.addAction(UIAlertAction(title: "3:5", style: .default) { (action) in
            self.setCropAspectRect(aspect: "3:5")
        })
        
        actionSheet.addAction(UIAlertAction(title: "3:4", style: .default) { (action) in
            self.setCropAspectRect(aspect: "3:4")
        })
        
        actionSheet.addAction(UIAlertAction(title: "5:7", style: .default) { (action) in
            self.setCropAspectRect(aspect: "5:7")
        })
        
        actionSheet.addAction(UIAlertAction(title: "9:16", style: .default) { (action) in
            self.setCropAspectRect(aspect: "9:16")
        })
        
        actionSheet.addAction(UIAlertAction(title: "Cancel", style: .cancel))
        
        present(actionSheet, animated: true, completion: nil)
    }
    
    @IBAction func onTouchLockAspectRatioButton(_ sender: UISwitch) {
        self.lockAspectRatio(sender.isOn)
    }
    
    //FIXME: Themes Preview
//    override open func customBorderColor() -> UIColor {
//        return UIColor.red
//    }
//
//    override open func customBorderWidth() -> CGFloat {
//        return 2.0
//    }
//
//    override open func customCornerBorderWidth() -> CGFloat {
//        return 4.0
//    }
//
//    override open func customCropLinesCount() -> Int {
//        return 3
//    }
//
//    override open func customGridLinesCount() -> Int {
//        return 4
//    }
//
//    override open func customCornerBorderLength() -> CGFloat {
//        return 30.0
//    }
//
//    override open func customIsHighlightMask() -> Bool {
//        return true
//    }
//
//    override open func customHighlightMaskAlphaValue() -> CGFloat {
//        return 0.3
//    }
    
    override open func customCanvasHeaderHeigth() -> CGFloat {
        var heigth: CGFloat = 0.0
        
        if UIDevice.current.orientation.isLandscape {
            heigth = 40.0
        } else {
            heigth = 100.0
        }
        
        return heigth
    }
}

extension ExampleCropViewController: HorizontalDialDelegate {
    func horizontalDialDidValueChanged(_ horizontalDial: HorizontalDial) {
        let degrees = horizontalDial.value
        let radians = IGRRadianAngle.toRadians(CGFloat(degrees))
        
        self.setupAngleLabelValue(radians: radians)
        self.changeAngle(radians: radians)
    }
    
    func horizontalDialDidEndScroll(_ horizontalDial: HorizontalDial) {
        self.stopChangeAngle()
    }
}
